### PR TITLE
[9.2] Provide a theme through an app

### PR DESCRIPTION
**Pull request by phisch**
## Description

This allows users to install apps that provide a theme. If an app enables itself as app-theme in its `app.php`:

```
/** @var Theme\ThemeService $themeService */
$themeService = \OC::$server->query('ThemeService');
$themeService->setAppTheme('theme-owncloud');
```

the app will be used as theme. If no app claims to be a theme, the existing logic will continue to work. That means, OC will look for the `theme` key in `config/config.php` and will use it. OC will still look for `themes/default` if no theme is configured.
## Related Issue
#20200
# Breaking changes

The documentation (https://doc.owncloud.org/server/9.0/developer_manual/core/theming.html) describes, that modifying an apps translation files would require a file in `theme/apps/app-name/l10n/language.js`. Modifying an apps template on the other hand would require a file in `theme/app-name/templates/template.php`.

This inconsistency has been removed. Templates will be located just like translations under `theme/apps/app-name/template/...`.

Theming an App now requires the corresponding files to be located in an "app" folder.

@DeepDiver1975, @PVince81, @butonic: what do you think about getting rid of this unnecessary complexity?
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
conflicting - please rebase - THX @PhilippSchaffrath

**Comment by PhilippSchaffrath** [comment]
The tests fail because vfsStream is missing. Does jenkins install require-dev composer dependencies?

**Comment by DeepDiver1975** [comment]
I'm the makefile there are targets to execute the tests. I guess there is composer call missing to install development dependencies

**Comment by butonic** [comment]
> @DeepDiver1975, @PVince81, @butonic: what do you think about getting rid of this unnecessary complexity?

+1

**Comment by PhilippSchaffrath** [comment]
Example Theme can be found here: https://github.com/PhilippSchaffrath/oc-app-theme

Just clone it into your apps directory and enable the app.

modified the following things as an example:
- login background colors
- header backround color
- file app sidebar and file list templates
- application title
- german translation for file "download"

**Comment by PVince81** [comment]
> I'm the makefile there are targets to execute the tests. I guess there is composer call missing to install development dependencies

The call is already there. Maybe Jenkins needs a quick "make clean".

If "make test-php" doesn't re-fetch the dev deps, it means that on this test env for some reason the timestamp of "lib/composer" is newer than the timestamp of "composer.json" or "composer.lock".

Maybe bumping the timestamp of composer.json might work ?

**Comment by PVince81** [comment]
Looks like it did run the command:
```
+ make test-php TEST_DATABASE=sqlite
php build/composer.phar install --dev
You are using the deprecated option "dev". Dev packages are installed by default now.
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.
Nothing to install or update
```

**Comment by butonic** [comment]
IMO we should introduce and use a new app type 'theme'. Then we don't need an app.php for a theme but can use 
```
	<types>
		<theme/>
	</types>
```
in the info.xml

- [ ] find the right place where we should load the `theme` apps
- [ ] add `theme` as a valid filetype

**Comment by PhilippSchaffrath** [comment]
I thought about exactly that too! It's on my todo now!

**Comment by PhilippSchaffrath** [comment]
Last commit checks if the app to load as the `theme` type, if so it registers it as such in the ThemeService. This makes the `app.php` obsolete. This works like @butonic described it.

Although that's just a small change, i am having a lot of trouble testing it because of the static nature of this class. Do you guys have a smart idea on how to test the last commit?

**Comment by DeepDiver1975** [comment]
Testing is not really possible with this legacy class.

**Comment by PVince81** [comment]
What happens if multiple theme apps are enabled at the same time ? The last one alphabetically will win ?

**Comment by PVince81** [comment]
> Theming an App now requires the corresponding files to be located in an "app" folder.
> what do you think about getting rid of this unnecessary complexity?

Which complexity exactly ?

Will this use case still work with your app ? => https://github.com/owncloud/apps/pull/2129

**Review by DeepDiver1975 — COMMENTED** [review]
> What happens if multiple theme apps are enabled at the same time ? The last one alphabetically will win ?

exactly

File: lib/private/Theme/ThemeService.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace OC\Theme;
+
+class ThemeService {
```
Comment (DeepDiver1975) [inline lib/private/Theme/ThemeService.php]: not needed anymore - theme is define via info.xml

**Comment by DeepDiver1975** [comment]
> Will this use case still work with your app ? => owncloud/apps#2129

I guess not

**Comment by DeepDiver1975** [comment]
You like super heros? Test this theme - https://github.com/DeepDiver1975/marvel-theme

**Comment by butonic** [comment]
@PhilippSchaffrath can you rebase? I seem to run into https://github.com/owncloud/core/issues/26401 when trying to upgrade to this branch. Has been fixed in https://github.com/owncloud/core/pull/26430

**Comment by PhilippSchaffrath** [comment]
Since the rebase somehow messed this branch up big time, i figured creating a new branch/pr by cherry-picking would be faster then trying to fix this branch.

New PR: #26462

**Comment by DeepDiver1975** [comment]
@PhilippSchaffrath please delete the branch as soon as you no longer need it - THX

**Comment by PhilippSchaffrath** [comment]
@DeepDiver1975 will do that as soon as #26462 checks are green, can you re-run the scruntinizer inspections please? It seems like they failed for no reason.

**Pull request closed without merging**